trie/snap_diskcache.go

```diff
@@ -0,0 +1,204 @@
+	cache.seq += 1
+
+	var size int64
+	for storage, node := range nodes {
+		if orig, exist := cache.nodes[storage]; !exist {
+			size += int64(node.memorySize(len(storage)))
+		} else {
+			size += int64(node.size) - int64(orig.size)
+			triedbGCNodesMeter.Mark(1)
+			triedbGCSizeMeter.Mark(int64(orig.size))
+		}
+		cache.nodes[storage] = node
+	}
+	if final := int64(cache.size) + size; final < 0 {
+		log.Error("Negative disk cache size", "previous", common.StorageSize(cache.size), "diff", common.StorageSize(size))
+		cache.size = 0
+	} else {
+		cache.size = uint64(final)
+	}
+	return cache
+}
+
+// revert applies the reverse diff to the local dirty node set.
+func (cache *diskcache) revert(diff *reverseDiff) error {
+	if cache.seq == 0 {
+		return errStateUnrecoverable
+	}
+	cache.seq -= 1
+
+	// If all the embedded state transitions are reverted,
+	// reset the cache entirely.
+	if cache.seq == 0 {
+		cache.reset()
+		return nil
+	}
+	for _, state := range diff.States {
+		_, ok := cache.nodes[string(state.Key)]
+		if !ok {
+			// TODO it should never happen, perhaps panic here.
+			owner, path := DecodeStorageKey(state.Key)
+			return fmt.Errorf("non-existent node (%x %v)", owner, path)
+		}
+		if len(state.Val) == 0 {
+			cache.nodes[string(state.Key)] = &cachedNode{
+				node: nil,
+				size: 0,
+				hash: common.Hash{},
+			}
+		} else {
+			cache.nodes[string(state.Key)] = &cachedNode{
+				node: rawNode(state.Val),
+				size: uint16(len(state.Val)),
+				hash: crypto.Keccak256Hash(state.Val),
+			}
+		}
+	}
+	return nil
+}
+
+// reset cleans up the disk cache.
+func (cache *diskcache) reset() {
+	cache.seq = 0
+	cache.nodes = make(map[string]*cachedNode)
+	cache.size = 0
+}
+
+// empty returns an indicator if diskcache contains any state transition inside.
+func (cache *diskcache) empty() bool {
+	return cache.seq == 0
+}
+
+// forEach iterates all the cached nodes and applies the given callback on them
+func (cache *diskcache) forEach(callback func(key string, node *cachedNode)) {
+	for storage, n := range cache.nodes {
+		callback(storage, n)
+	}
+}
+
+// mayFlush persists the in-memory dirty trie node into the disk if the predefined
+// memory threshold is reached. Note, all data must be written to disk atomically.
+// This function should never be called simultaneously with other map accessors.
+func (cache *diskcache) mayFlush(db ethdb.KeyValueStore, clean *fastcache.Cache, diffid uint64, force bool) error {
+	if cache.size <= cacheSizeLimit && !force {
+		return nil
+	}
+	// Ensure the given reverse diff identifier is aligned
+	// with the internal counter.
+	head := rawdb.ReadReverseDiffHead(db)
+	if head+cache.seq != diffid {
+		return errors.New("invalid reverse diff id")
+	}
+	var (
+		start = time.Now()
+		batch = db.NewBatchWithSize(int(cacheSizeLimit))
+	)
+	for storage, n := range cache.nodes {
+		if n.node == nil {
+			rawdb.DeleteTrieNode(batch, []byte(storage))
```

Comment:
Shouldn't we also remove it from the `clean` cache? Otherwise an old stale data recently accessed might be returned.

Reply:
Ooooh, this is hash to node mapping, **not** path to node, nevermind!